src/librustc_mir/transform/const_prop.rs

```diff
@@ -370,10 +405,34 @@ impl<'a, 'mir, 'tcx> ConstPropagator<'a, 'mir, 'tcx> {
                     this.ecx.cast(op, kind, dest.into())?;
                     Ok(dest.into())
                 })
-            }
+            },
+            Rvalue::Len(ref place) => {
+                let place = self.eval_place(&place, source_info)?;
+                let mplace = place.to_mem_place();
 
-            // FIXME(oli-obk): evaluate static/constant slice lengths
-            Rvalue::Len(_) => None,
+                let is_slice =
+                    if let ty::Slice(_) = mplace.layout.ty.sty { true }
```

Comment:
This is scary. Do we really encounter both `&[T]` and `[T]`?

Reply:
Yeah, after `*&[T]` is evaluated, we're left with a bare `[T]`.